src/libraries/System.Text.Encoding/tests/Encoding/TranscodingStreamTests.cs

```diff
@@ -452,6 +470,56 @@ public async Task ReadAsync_Memory(int bufferLength)
             });
         }
 
+        [Fact]
+        public async Task ReadAsync_LoopsWhenPartialDataReceived()
+        {
+            // Validates that the TranscodingStream will loop instead of returning 0
+            // if the inner stream read partial data and GetBytes cannot make forward progress.
+
+            using AsyncComms comms = new AsyncComms();
+            Stream transcodingStream = Encoding.CreateTranscodingStream(comms.ReadStream, Encoding.UTF8, Encoding.UTF8);
+
+            // First, ensure that writing [ C0 ] (always invalid UTF-8) to the stream
+            // causes the reader to return immediately with fallback behavior.
+
+            byte[] readBuffer = new byte[1024];
+            comms.WriteBytes(new byte[] { 0xC0 });
+
+            int numBytesRead = await transcodingStream.ReadAsync(readBuffer.AsMemory());
+            Assert.Equal(new byte[] { 0xEF, 0xBF, 0xBD }, readBuffer[0..numBytesRead]); // fallback substitution
+
+            // Next, ensure that writing [ C2 ] (partial UTF-8, needs more data) to the stream
+            // causes the reader to asynchronously loop, returning "not yet complete".
+
+            readBuffer = new byte[1024];
+            comms.WriteBytes(new byte[] { 0xC2 });
+
+            ValueTask<int> task = transcodingStream.ReadAsync(readBuffer.AsMemory());
+            Assert.False(task.IsCompleted);
+            comms.WriteBytes(new byte[] { 0x80 }); // [ C2 80 ] is valid UTF-8
+
+            numBytesRead = await task; // should complete successfully
+            Assert.Equal(new byte[] { 0xC2, 0x80 }, readBuffer[0..numBytesRead]);
+
+            // Finally, ensure that writing [ C2 ] (partial UTF-8, needs more data) to the stream
+            // followed by EOF causes the reader to perform substitution before returning EOF.
+
+            readBuffer = new byte[1024];
+            comms.WriteBytes(new byte[] { 0xC2 });
+
+            task = transcodingStream.ReadAsync(readBuffer.AsMemory());
```

Comment:
Would it make sense to add a test that validates behaviour when passing a cancellation token in this scenario?

Reply:
That's already part of the main `ReadAsync` test logic. See below.

https://github.com/dotnet/runtime/blob/b1a417875e68e1961c163c1f0611549c2c2e04f6/src/libraries/System.Text.Encoding/tests/Encoding/TranscodingStreamTests.cs#L487-L504